pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -0,0 +1,327 @@
+	e := &DockercfgController{
+		client: cl,
+	}
+
+	_, e.serviceAccountController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return e.client.ServiceAccounts(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return e.client.ServiceAccounts(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.ServiceAccount{},
+		options.Resync,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    e.serviceAccountAdded,
+			UpdateFunc: e.serviceAccountUpdated,
+		},
+	)
+
+	e.dockerURL = OpenshiftDockerURL
+
+	return e
+}
+
+// DockercfgController manages dockercfg secrets for ServiceAccount objects
+type DockercfgController struct {
+	stopChan chan struct{}
+
+	client client.Interface
+
+	dockerURL     string
+	dockerURLLock sync.Mutex
+
+	serviceAccountController *framework.Controller
+}
+
+// Runs controller loops and returns immediately
+func (e *DockercfgController) Run() {
+	if e.stopChan == nil {
+		e.stopChan = make(chan struct{})
+		go e.serviceAccountController.Run(e.stopChan)
+	}
+}
+
+// Stop gracefully shuts down this controller
+func (e *DockercfgController) Stop() {
+	if e.stopChan != nil {
+		close(e.stopChan)
+		e.stopChan = nil
+	}
+}
+
+func (e *DockercfgController) SetDockerURL(newDockerURL string) {
+	e.dockerURLLock.Lock()
+	defer e.dockerURLLock.Unlock()
+
+	e.dockerURL = newDockerURL
+}
+
+// serviceAccountAdded reacts to a ServiceAccount creation by creating a corresponding ServiceAccountToken Secret
+func (e *DockercfgController) serviceAccountAdded(obj interface{}) {
+	serviceAccount := obj.(*api.ServiceAccount)
+
+	if err := e.createDockercfgSecretIfNeeded(serviceAccount); err != nil {
+		glog.Error(err)
+	}
+}
+
+// serviceAccountUpdated reacts to a ServiceAccount update (or re-list) by ensuring a corresponding ServiceAccountToken Secret exists
+func (e *DockercfgController) serviceAccountUpdated(oldObj interface{}, newObj interface{}) {
+	newServiceAccount := newObj.(*api.ServiceAccount)
+
+	if err := e.createDockercfgSecretIfNeeded(newServiceAccount); err != nil {
+		glog.Error(err)
+	}
+}
+
+func getDockercfgSecretNamePrefix(serviceAccount *api.ServiceAccount) string {
+	return serviceAccount.Name + "-dockercfg-"
+}
+func getTokenSecretNamePrefix(serviceAccount *api.ServiceAccount) string {
+	return serviceAccount.Name + "-token-"
+}
+
+// createDockercfgSecretIfNeeded makes sure at least one ServiceAccountToken secret exists, and is included in the serviceAccount's Secrets list
+func (e *DockercfgController) createDockercfgSecretIfNeeded(serviceAccount *api.ServiceAccount) error {
+
+	// look for an ImagePullSecret in the form
+	dockercfgSecretName := ""
+	foundDockercfgImagePullSecret := false
+	for _, pullSecret := range serviceAccount.ImagePullSecrets {
+		if strings.HasPrefix(pullSecret.Name, getDockercfgSecretNamePrefix(serviceAccount)) {
```

Comment:
This could be improved if secret.Strategy had it's own GenerateName func that used more of the available name length, but would still be an issue for service accounts at or near the maximum name length.